samtranslator/model/sam_resources.py

```diff
@@ -2407,3 +2420,206 @@ def _set_resolver_code_settings(resolver_code_settings: aws_serverless_graphqlap
+                attributes=self.resource_attributes,
+            )
+
+            # "Id" refers to the "LogicalId" attribute for a "AppSync::FunctionConfiguration" resource.
+            # "Id" is a mutually exclusive property to every other property. If this function has it
+            # defined, then we make sure it's the only property, and continue to next function.
+            if function.Id:
+                keys = remove_none_values(function.dict()).keys()  # remove undefined properties, then get keys
+                if len(keys) != 1:
+                    raise InvalidResourceException(
+                        relative_id, "'Id' cannot be defined with other properties in Function."
+                    )
+                continue
+
+            # TODO: Make utility function that handles these mutual exclusive errors for us
+            if not function.DataSource and not function.DataSourceName:
+                raise InvalidResourceException(relative_id, "One of 'DataSource' or 'DataSourceName' must be set.")
+
+            if function.DataSource and function.DataSourceName:
+                raise InvalidResourceException(
+                    relative_id, "Both 'DataSourceType' and 'DataSourceName' cannot be defined at the same time."
+                )
+
+            func_config.ApiId = api_id
+            func_config.Name = function.Name or relative_id
+            func_config.Code, func_config.CodeS3Location = self._parse_code_properties(
+                function, resolver_code_settings, func_config.Name, relative_id
+            )
+            func_config.DataSourceName = self._parse_datasource_name(
+                relative_id, function, datasources, none_datasource_logical_id
+            )
+            func_config.MaxBatchSize = passthrough_value(function.MaxBatchSize)
+            func_config.Description = passthrough_value(function.Description)
+            func_config.Runtime = self._parse_runtime(function, resolver_code_settings, relative_id)
+
+            if function.Sync:
+                func_config.SyncConfig = self._parse_function_sync(function.Sync)
+
+            resources.append(func_config)
+
+        return resources
+
+    def _check_and_construct_none_datasource(
+        self,
+        functions: Dict[str, aws_serverless_graphqlapi.Function],
+        api_id: Intrinsicable[str],
+        none_datasource_logical_id: str,
+    ) -> Optional[DataSource]:
+        """
+        Checks if a DataSource with type "NONE" needs to be created.
+
+        Within a Serverless::GraphQLApi Function resource, customers can create a
+        DataSource of Type "NONE" for quick use in Functions. To do so, a customer
+        can input "NONE", "None", or "none" in the "DataSource" property. Only one
+        DataSource will be created for each GraphQLApi, and all GraphQLApi functions
+        with this none input will reference it.
+        """
+        for function in functions.values():
```

Comment:
As a suggestion: 
no need to loop through functions 1 more time, if we can in O(1) check if NoneDataSource already exists or not and either use it or create a new one. Then it can be done for each function in `_parse_datasource_name()`